HDFBatchEventsOutputer.cc

```diff
@@ -0,0 +1,356 @@
+  std::transform(iEventIDs.begin(), iEventIDs.end(), std::back_inserter(ids), [](auto const&id) {return id.event;});
+  //std::cout <<" ids "<<ids.size()<<std::endl;
+  write_ds<unsigned long long>(group_, EVENTS_DSNAME, ids);
+  //std::cout <<"wrote ids"<<std::endl;
+  write_ds<char>(group_, PRODUCTS_DSNAME, iBuffer);
+  write_ds<uint32_t>(group_, OFFSETS_DSNAME, iOffsets); 
+}
+
+void 
+HDFBatchEventsOutputer::writeFileHeader(SerializeStrategy const& iSerializers) {
+  constexpr hsize_t ndims = 1;
+  constexpr hsize_t dims[ndims] = {0};
+  hsize_t chunk_dims[ndims] = {static_cast<hsize_t>(chunkSize_)};
+  constexpr hsize_t  max_dims[ndims] = {H5S_UNLIMITED};
+  auto space = hdf5::Dataspace::create_simple (ndims, dims, max_dims); 
+  auto prop   = hdf5::Property::create();
+  prop.set_chunk(ndims, chunk_dims);
+  hdf5::Dataset::create<int>(group_, EVENTS_DSNAME, space, prop);
+  hdf5::Dataset::create<char>(group_, PRODUCTS_DSNAME, space, prop);
+  hdf5::Dataset::create<int>(group_, OFFSETS_DSNAME, space, prop);
+
+  const auto scalar_space  = hdf5::Dataspace::create_scalar();
+  hdf5::Attribute::create<int>(group_, RUN_ANAME, scalar_space);
+  hdf5::Attribute::create<int>(group_, LUMISEC_ANAME, scalar_space);
+  hdf5::Attribute::create<int>(group_, COMPRESSION_LEVEL_ANAME, scalar_space);
+  hdf5::Attribute::create<int>(group_, COMPRESSION_CHOICE_ANAME, scalar_space);
+  constexpr hsize_t     str_dims[ndims] = {10};
+  auto const attr_type = H5Tcopy (H5T_C_S1);
+  H5Tset_size(attr_type, H5T_VARIABLE);
+  auto const attr_space  = H5Screate(H5S_SCALAR);
+  hdf5::Attribute compression = hdf5::Attribute::create<std::string>(group_,COMPRESSION_ANAME, attr_space); 
+  for(auto const& s: iSerializers) {
+    std::string const type(s.className());
+    std::string const name(s.name());
+    hdf5::Attribute prod_name = hdf5::Attribute::create<std::string>(group_, name.c_str(), attr_space); 
+    prod_name.write<std::string>(type);
+  }
+}
+
+std::pair<std::vector<uint32_t>, std::vector<char>> HDFBatchEventsOutputer::writeDataProductsToOutputBuffer(SerializeStrategy const& iSerializers) const{
+  //Calculate buffer size needed
+  uint32_t bufferSize = 0;
+  std::vector<uint32_t> offsets;
+  offsets.reserve(iSerializers.size()+1);
+  for(auto const& s: iSerializers) {
+    auto const blobSize = s.blob().size();
+    offsets.push_back(bufferSize);
+    bufferSize += blobSize;
+  }
+  offsets.push_back(bufferSize);
+
+  //initialize with 0
+  std::vector<char> buffer(bufferSize, 0);
+  
+  {
+    uint32_t index = 0;
+    for(auto const& s: iSerializers) {
+      auto offset = offsets[index++];
+      std::copy(s.blob().begin(), s.blob().end(), buffer.begin()+offset );
+    }
+    assert(buffer.size() == offsets[index]);
+  }
+
+  if(compressionChoice_ == CompressionChoice::kEvents or compressionChoice_ == CompressionChoice::kBoth) {
+    auto cBuffer  = pds::compressBuffer(0,0, compression_, compressionLevel_, buffer);
+
+    return {offsets, cBuffer};
```

Comment:
I believe an additional copy of `offsets` and `cBuffer` is occurring here (at least my tests on quick-bench.com suggest this).  Here and below I recommend:

```diff
-  return {offsets, cBuffer};
+  return {move(offsets), move(cBuffer)};
```

Reply:
done